### DISCUSSION TITLE
Support other git hosting platforms (GitLab and Gitea)?

### DISCUSSION THREAD

**Discussion by shenxianpeng** [cpp-linter/cpp-linter#127] [Ideas]
The user in [this issue](https://github.com/cpp-linter/cpp-linter-action/issues/281) asked if we support platforms like GitLab or Gitea. 

The short answer is NO; we don’t currently support other platforms directly.

However, for GitLab and Gitea, you can use [cpp-linter/cpp-linter-hooks](https://github.com/cpp-linter/cpp-linter-hooks) with a `.pre-commit-config.yaml` file in your CI pipeline.

If you’d like us to add direct support for these platforms, please let us know—this helps us gauge the level of interest.

**Comment by 2bndy5** [comment]
I'd prefer a separate issue/thread for each requested git server; 1 separate issue for GitLab (see #101) and 1 separate issue for Gitea.

**Comment by shenxianpeng** [comment]
Maybe we can move this issue to discussion for collection feedback and then create separate issues for GitLab and Gitea if needed.

**Comment by shenxianpeng** [comment]
## feat: cross-CI support (GitLab CI, Bitbucket Pipelines)

### Summary
Make cpp-linter platform-agnostic so it's a first-class citizen on GitLab CI and
Bitbucket Pipelines, not just GitHub Actions. Reframe the project from "a GitHub
Action" into "a cross-platform C/C++ quality gate."

### Motivation
Competing C/C++ linting tools advertise support across GitHub / GitLab / Bitbucket
/ Azure DevOps, while cpp-linter is currently GitHub + Debian-Linux centric. GitLab
and Bitbucket each have their own CI (`.gitlab-ci.yml`, `bitbucket-pipelines.yml`)
where users already run shell commands — so the gap isn't "can it run" but "does it
feel native" (results on the MR/PR, diff-aware, zero config).

### Design principle
Keep all platform differences inside the **CLI**, not in per-platform integrations.
Platforms only need to (a) run the CLI and (b) consume its output. This keeps
maintenance linear: one CLI + a thin template/pipe per platform.

### Proposed work
**CLI (the bulk of the effort):**
- [ ] Add output formats: SARIF (tracked separately) and **CodeClimate JSON** (for GitLab Code Quality).
- [ ] Auto-detect CI environment (`GITHUB_*` / `CI_*` / `BITBUCKET_*`) for PR/MR number, diff, and tokens.
- [ ] Unified diff-aware mode: lint only changed files on PR/MR, full tree on default branch, on every platform.

**GitLab (pilot deliverable):**
- [ ] Maintained `.gitlab-ci.yml` template (and/or CI component).
- [ ] Emit CodeClimate report via `artifacts:reports:codequality` so findings show inline on MRs.

**Bitbucket:**
- [ ] A Bitbucket Pipe (docker image + `pipe.yml`) that runs the CLI and uploads results via the Code Insights API (Reports + Annotations).

**Later (optional):**
- [ ] Azure DevOps task.

### Acceptance criteria
- [ ] Same CLI produces inline findings on GitHub PRs, GitLab MRs, and Bitbucket PRs.
- [ ] Diff-aware verified on all three platforms (changed files only on PR/MR).
- [ ] Copy-paste config example documented for each platform.

### Out of scope
Local pre-commit usage is already covered by `cpp-linter-hooks`; this issue is CI only.
